File: Homework/Quang_Minh/HomeWorkLesson7/BinaryTreePreorderTraversal.java
package HomeWorkLesson7;

import java.util.ArrayList;
import java.util.List;

public class BinaryTreePreorderTraversal {
    List<Integer> result = new ArrayList<>();
    public List<Integer> preorderTraversal(TreeNode root) {
        dfs(root);
        return result;
    }
    public void dfs(TreeNode node){
        if(node == null) return;
        result.add(node.val);
        dfs(node.left);
        dfs(node.right);
    }
}

File: Homework/Van_Le/lesson7/BinaryTreePreorderTraversal/BinaryTreePreorderTraversal.cs
namespace LeetCodeSolutions
{
    public class BinaryTreePreorderTraversal
    {
        public IList<int> PreorderTraversal(TreeNode root)
        {
            IList<int> output = new List<int>();

            TreeNode node = root;
            while (node != null)
            {
                if (node.left == null)
                {
                    output.Add(node.val);
                    node = node.right;
                }
                else
                {
                    TreeNode predecessor = node.left;
                    while ((predecessor.right != null) && (predecessor.right != node))
                    {
                        predecessor = predecessor.right;
                    }

                    if (predecessor.right == null)
                    {
                        output.Add(node.val);
                        predecessor.right = node;
                        node = node.left;
                    }
                    else
                    {
                        predecessor.right = null;
                        node = node.right;
                    }
                }
            }
            return output;
        }
    }
}
